### PR TITLE
CosmosChainProcessor Client message handlers

### PR DESCRIPTION
**Pull request by agouin**
Adds `latestClientState` caching on the `CosmosChainProcessor`, which will be used in a future PR for constructing trusted IBC headers so that `MsgUpdateClient` messages can be constructed for the counterparty chain.

Changes cache of data shared between `ChainProcessor` and `PathProcessor` from `ChannelMessageCache` to `IBCMessagesCache` to account for not only packet flow messages, but also connection and channel handshake messages, which the `PathProcessor` can handle, if applicable, to complete the connection and channel handshakes (in a future PR).

Adds `Merge` and `Retain` helpers for the different cached data types.

### CODE REVIEW

File: relayer/chains/cosmos/types.go [hunk 1]
```diff
@@ -67,6 +76,27 @@ type clientInfo struct {
 	header          []byte
 }
 
+type latestClientState map[string]clientInfo
```
Comment (mark-rushakoff) [inline relayer/chains/cosmos/types.go:80]: It's nice to have a comment noting what the key represents -- client ID? chain ID with client ID?

File: relayer/processor/types.go [hunk 2]
```diff
@@ -68,75 +94,67 @@ func (channelKey ChannelKey) Counterparty() ChannelKey {
...
-
-// Clone will create a deep copy of a MessageCache.
-func (c MessageCache) Clone() MessageCache {
-	newMessageCache := make(MessageCache)
+// Clone will create a deep copy of a PacketMessagesCache.
+func (c PacketMessagesCache) Clone() PacketMessagesCache {
+	newPacketMessagesCache := make(PacketMessagesCache)
 	for mk, mv := range c {
-		newSequenceCache := make(SequenceCache)
+		newPacketSequenceCache := make(PacketSequenceCache)
```
Comment (mark-rushakoff) [inline relayer/processor/types.go]: Slightly better if these `make` calls can use size hints to avoid regrowing the maps.

**Review by mark-rushakoff — APPROVED** [review]
Changes all look straightforward here AFAICS.

File: relayer/chains/cosmos/message_handlers.go [hunk 3]
```diff
@@ -123,28 +129,67 @@ func (ccp *CosmosChainProcessor) handleMsgTimeoutOnClose(p MsgHandlerParams) boo
...
 		zap.String("dst_port", pi.packet.DestinationPort),
 	}
 	fields = append(fields, additionalFields...)
-	ccp.log.Debug("Observed IBC message", fields...)
+	ccp.logObservedIBCMessage(message, fields...)
 }
 
 // END packet msg handlers
+
+// BEGIN client msg handlers
```
Comment (DavidNix) [inline relayer/chains/cosmos/message_handlers.go]: nit:

These look like pragma comments as common in other languages. Perhaps your IDE has special treatment of these comments?

In Go, I typically use (and see others use) the pattern of different files for each area of responsibility. 

E.g. in this case I'd probably create a `client_msg_handlers.go` file.

**Review by DavidNix — APPROVED** [review]

File: relayer/chains/cosmos/types.go (hunk 1 above)
Comment (agouin) [reply relayer/chains/cosmos/types.go:80]: done

File: relayer/processor/types.go (hunk 2 above)
Comment (agouin) [reply relayer/processor/types.go]: done

File: relayer/chains/cosmos/message_handlers.go (hunk 3 above)
Comment (agouin) [reply relayer/chains/cosmos/message_handlers.go]: done

**Pull request merged**